src/hotspot/cpu/s390/assembler_s390.inline.hpp

```diff
@@ -30,30 +30,62 @@
 #include "asm/codeBuffer.hpp"
 #include "code/codeCache.hpp"
 
-// Convention: Use Z_R0 and Z_R1 instead of Z_scratch_* in all
-// assembler_s390.* files.
 
-// Local implementation of byte emitters to help inlining.
+inline unsigned int Assembler::emit_instruction(unsigned long x, unsigned int len) {
+  CodeSection*  cs = code_section();
+  address      pos = pc();
+  assert(((intptr_t)pos & 0x01L) == 0, "instruction address alignment error");
+  if (len == 0) {
+    len = instr_len(x);
+  }
+
+  switch (len) {
+    case 2:
+      *(unsigned short*)(pos)     = (unsigned short)x;
+      break;
+    case 4:
+      if (((intptr_t)pos & 0x03) == 0) {
+        *(unsigned int*)(pos)     = (unsigned int)x;
+      } else {
+        *(unsigned short*)(pos)   = (unsigned short)(x>>16);
+        *(unsigned short*)(pos+2) = (unsigned short)x;
+      }
```

Comment:
Looks correct, but what is the motivation behind splitting unaligned stores? z can store to unaligned memory.

Reply:
Will go back to single int store for 4-byte instructions.

Reply:
Ok. I'm fine with either version. Maybe the reason you did it this way was to be more consistent with the 6 Byte version.